### DISCUSSION TITLE
Week 9: Assignment + office hours

### DISCUSSION THREAD

**Discussion by hamzamajoka** [gui2de/ppol768-spring23#300] [Assignments Feedback]
I've uploaded the code for multi-level data simulation on Github + uploaded the lecture recording on Box. 

I'll be holding office hours in gui2de office today at 3 pm. You can also post your questions here if you can't make it to the office hours.

**Comment by mikegiunti** [comment]
Hello, I am not progressing with this assignment at all, since there is no point of reference found online or in here for it.

My code so far is this. I don't know if it is correct, and I am not sure how I would collect the data of five different regressions in a way that is not redundant (copying an pasting the program), or how to replicate the visualization you made in class. 


clear 
set obs `samplesize'

gen region = _n
gen region_effect = rnormal(0,2)
expand 200+int((250-200+1)*runiform())

bysort region : gen artisan = _n
gen artisan_effect = rnormal(0,50)


gen business_exp = rnormal()
gen sales = rnormal()
gen customers = rnormal()

gen business_success = 3 + (-1)*region + .5*business_exp + 6*customers + artisan_effect + region_effect
gen treat_payment = region + business_exp + sales

reg business_success treat_payment
reg business_success treat_payment i.region
reg business_success treat_payment i.region business_exp
reg business_success treat_payment i.region business_exp sales
reg business_success treat_payment i.region business_exp customers
reg business_success treat_payment i.region business_exp sales customers
reg business_success treat_payment business_exp sales customers

I'm sure I would understand these concepts more if I could actually execute the code, and any hint into actually progressing would be great, even just samples or anything of the sort.

**Reply by hamzamajoka** [reply]
Thanks for posting it here. I've made some changes to your code, let me know if it makes sense:

```
local samplesize 5 //ah1152: sample size in your code is actually the number of strata 

clear
set obs `samplesize'

gen region = _n 
gen region_effect = rnormal(0,2) //generating 
expand 200 //ah1152: making uniform size strata to keep things simple, you can revert it back once your code us working

bysort region : gen artisan = _n
gen artisan_effect = rnormal(0,5)  //changing the stand_dev from 50 to 5, 

gen business_exp = rnormal()
gen sales = rnormal()
gen customers = rnormal()

*generate treatment dummy variable which is influeced by region (strata) and business_exp and sales
gen treat = (region/(5) + business_exp + sales + rnormal()) > 0 

*generate dependent variable, where treatment variable has some impact(say 2.5) 
gen business_success = 3 + (-1)*region + .5*business_exp + 6*customers + artisan_effect + region_effect + 2.5*treat



reg business_success treat
reg business_success treat i.region
reg business_success treat i.region business_exp 
reg business_success treat i.region business_exp sales
reg business_success treat i.region business_exp customers
reg business_success treat i.region business_exp sales customers
reg business_success treat business_exp sales customers

//add code to return scalars for the coefficient of treatment from each model so that you can store the results when you use simulate command
```

The basic idea now is that you know from the DGP that treatment is increasing the business_success by 2.5 units. The correct model should give you a beta for treat close to 2.5.

**Reply by hamzamajoka** [reply]
I've further simplified your code to create a Stata program.

```
capture program drop reg_conf
program define reg_conf , rclass
syntax, strata(integer) obs_in_strata(integer)


clear
set obs `strata'

gen region = _n 
expand `obs_in_strata' //ah1152: making uniform size strata to keep things simple, you can revert it back once your code us working

*I removed region and artisan effects to simplify the model, I've instead added random noise in DGP

gen business_exp = rnormal()
gen sales = rnormal()
gen customers = rnormal()

*generate treatment dummy variable which is influeced by region (strata) and business_exp and sales
gen treat = (region/(5) + business_exp + sales + rnormal()) > 0 

*generate dependent variable, where treatment variable has some impact(say 2.5 units) 
gen business_success = 3 + (-1)*region + business_exp + 6*customers + 2.5*treat + 2*rnormal()



reg business_success treat
	return scalar model_1=_b[treat]
	
reg business_success treat i.region
	return scalar model_2=_b[treat]

reg business_success treat i.region business_exp 
	return scalar model_3=_b[treat]

reg business_success treat i.region customers sales
	return scalar model_4=_b[treat]

reg business_success treat  business_exp sales customers
	return scalar model_5=_b[treat]
end

simulate model_1=r(model_1) model_2=r(model_2) ///
           model_3=r(model_3) model_4=r(model_4) ///
           model_5=r(model_5)  ///
  , r(150): reg_conf, strata(5) obs_in_strata(500)
```

You can build on this program to re-add some of code I took out. You can then run the simulate command at different sample sizes and compare the _b[treat] of each model.

**Comment by pew36** [comment]
How can I avoid receiving an error message that's along the lines of, "file /var/folders/59/d5nx34g51c7ch04dtlqh4mh00000gn/T//S_37872.000001 already exists."

**Reply by hamzamajoka** [reply]
It's difficult to debug without looking at the code + underlying data. That being said, I'm like 51% sure that you are getting this message because you are saving a tempfile within a loop **without the REPLACE option**.

**Reply by pew36** [reply]
Thank you, Ali, for your help. Please find my pasted Stata output below. I'm still unable to run the simulate command within a loop, even after adding the REPLACE option (which was a helpful addition):

cd "/Users/peytonweber/Desktop/GitHub/ppol768-spring23/Individual Assignments/Weber Peyton/Week 09"
*Reviewer to change working directory accordingly 
*Clear before running code. 
clear all 

*Part 2: 

*Using capture program drop so that re-running the .do file is feasible. 
capture program drop RDI
*Define the program.
program define RDI, rclass
*Sample size will be the argument in the defined program. 
syntax, samplesize(integer) 
clear
*Set seed for replicability. 
set seed 04182023
*Create empty observations. 
set obs `samplesize'
*Create school observations, like in in-class exercise. 
gen school = _n
*Ali's notes say the following code is "school effects," but I don't recall what he means by this.
gen u_i = rnormal(0,2)
*Create new variables that randomly assign the observations to either be rural or ubran.
gen urban = runiform()<0.50
*Create ten classrooms in each school. 
expand 10
*Generate classroom IDs by sorting the school variable and then using the subsequent order of observations, _n. 
bysort school: generate classroom = _n 
*Classroom effects (?) 
gen u_ij = rnormal(0,3)
*Creating a variable for years of teaching experience. 
bysort school: generate teach_exp = 5+int((20-5+1)*runiform())
*Generate student-level dataset where each school-class will have 16-25 students.
expand 16+int((25-16+1)*runiform())
*Create student IDs
bysort school classroom: generate child = _n 
*Create student-level effects
generate e_ijk = rnormal(0,5)
*Let's now create a variable representing a given mother's level of education. 
generate temprand = runiform()
egen mother_educ = cut(temprand), at(0,0.5, 0.9, 1) icodes
*Create labeled categories of various levels of education. 
label define mother_educ 0 "HighSchool" 1 "College" 2 ">College"
*Create the treatment. 
gen treat = cond(teach_exp <= 10 & mother_educ <= 1, 1, 0)

gen z = 0.5*treat + rnormal()
*Replicating something we did in class.
generate score = 70 ///
        + 5*z ///
        + (-2)*urban ///
        + 1.5*teach_exp  ///
        + 0* mother_educ ///
        + u_i + u_ij + e_ijk
		
gen h = score + 3*treat + rnormal()

*First regression:
reg score treat
mat a = r(table)
return scalar Beta1 = a[1,1]

*Second regression:
reg score treat urban teach_exp mother_educ 
mat a = r(table)
return scalar Beta2 = a[1,1]

*Third regression: 
reg score treat h urban teach_exp mother_educ h 
mat a = r(table)
return scalar Beta3 = a[1,1]

*Fourth regression: 
reg score treat urban teach_exp mother_educ i.school
mat a = r(table)
return scalar Beta4 = a[1,1]

*Fifth regression: 
reg score treat urban teach_exp mother_educ i.(school classroom)
mat a = r(table)
return scalar Beta5 = a[1,1]

end

clear
*Defining a temporary space Stata to store data created in loop to follow:
tempfile combinedtwo
*Telling Stata not to present an error message that the tempfile has no data in it.
save `combinedtwo', replace emptyok

forvalues i=1/8 {
	local samplesize= 2^`i'
	display as error "iteration = `i'" 
	tempfile sims , replace
	simulate beta1 = r(Beta1) beta2 = r(Beta2) beta3 = r(Beta3) beta4 = r(Beta4) beta5 = r(Beta5) ///
	, reps(50) seed(4454) saving(`sims') ///
	: RDI, samplesize(`samplesize') 
	 
	use `sims', clear
	gen samplesize = `samplesize'
	append using `combinedtwo'
	save `combinedtwo'
}

When I last ran the code above, I receive an error message after "iteration = 1" that says, "_, invalid name."

**Reply by hamzamajoka** [reply]
Hi Peyton, 

There are a couple of issues with your code. It would be much more efficient if we can debug this over zoom. I'm free until 1:30 pm, email me the zoom link if it works for you. Thanks!

Ali

**Reply by hamzamajoka** [reply]
I made some changes to your code, you should be able to run it now:


```
*Clear before running code.
clear all

*Part 2:

*Using capture program drop so that re-running the .do file is feasible.
capture program drop RDI
*Define the program.
program define RDI, rclass
syntax, samplesize(integer)  //*Sample size will be the argument in the defined program.
clear 
*set seed 04182023 //Set seed for replicability.
*commenting out the seed part becuase setting the seed inside the program will generate the exact same output every single time
*Create empty observations.
set obs `samplesize'
*Create school observations, like in in-class exercise.
gen school = _n
*Ali's notes say the following code is "school effects," but I don't recall what he means by this.
gen u_i = rnormal(0,2)
*Create new variables that randomly assign the observations to either be rural or ubran.
gen urban = runiform()<0.50
*Create ten classrooms in each school.
expand 10
*Generate classroom IDs by sorting the school variable and then using the subsequent order of observations, _n.
bysort school: generate classroom = _n
*Classroom effects (?)
gen u_ij = rnormal(0,3)
*Creating a variable for years of teaching experience.
bysort school: generate teach_exp = 5+int((20-5+1)*runiform())
*Generate student-level dataset where each school-class will have 16-25 students.
expand 16+int((25-16+1)*runiform())
*Create student IDs
bysort school classroom: generate child = _n
*Create student-level effects
generate e_ijk = rnormal(0,5)
*Let's now create a variable representing a given mother's level of education.
generate temprand = runiform()
egen mother_educ = cut(temprand), at(0,0.5, 0.9, 1) icodes
*Create labeled categories of various levels of education.
label define mother_educ 0 "HighSchool" 1 "College" 2 ">College"
*Create the treatment.
gen treat = cond(teach_exp <= 10 & mother_educ <= 1, 1, 0)

gen z = 0.5*treat + rnormal()
*Replicating something we did in class.
generate score = 70 ///
+ 5*z ///
+ (-2)*urban ///
+ 1.5*teach_exp ///
+ mother_educ ///
+ u_i + u_ij + e_ijk

gen h = score + 3*treat + rnormal()

*First regression:
reg score treat
mat a = r(table)
return scalar Beta1 = a[1,1]

*Second regression:
reg score treat urban teach_exp mother_educ
mat a = r(table)
return scalar Beta2 = a[1,1]

*Third regression:
reg score treat h urban teach_exp mother_educ h
mat a = r(table)
return scalar Beta3 = a[1,1]

*Fourth regression:
reg score treat urban teach_exp mother_educ i.school
mat a = r(table)
return scalar Beta4 = a[1,1]

*Fifth regression:
reg score treat urban teach_exp mother_educ i.(school classroom)
mat a = r(table)
return scalar Beta5 = a[1,1]

end

clear
*Defining a temporary space Stata to store data created in loop to follow:
tempfile combinedtwo
*Telling Stata not to present an error message that the tempfile has no data in it.
save `combinedtwo', replace emptyok

forvalues i=1/8 {
local samplesize= 2^`i' 
display as error "iteration = `i'"
tempfile sims 
simulate beta1 = r(Beta1) beta2 = r(Beta2) beta3 = r(Beta3) beta4 = r(Beta4) beta5 = r(Beta5) ///
, reps(50) seed(4454) saving(`sims') /// 
: RDI, samplesize(`samplesize')

use `sims', clear
gen samplesize = `samplesize'
append using `combinedtwo'
save `combinedtwo', replace 
}

use `combinedtwo'
```

**Comment by mikegiunti** [comment]
Hello, this is more for week 10 feedback, but when the assingment says "Instead of having strata groups contributing to the main effect, create some portion of the random error term at the strata level (now, they are clusters, rather than strata). Use a moderately large number of clusters, and also assign treatment at the cluster level." do they mean something like the following code?

```
clear
set obs `strata'

gen region = _n 
gen region_effect = rnormal(0,2)
gen treat = (region + rnormal()) > 0 

expand `lower_obs'+int((`upper_obs'-`lower_obs'+1)*runiform())

bysort region : gen artisan = _n
gen artisan_effect = rnormal(0,5)

gen business_exp = rnormal()
gen sales = rnormal()
gen customers = rnormal()


gen business_success = 3 + (-1)*region + business_exp + 6*customers + rnormal()*treat + region_effect + artisan_effect
```

**Comment by genjuro86** [comment]
Hello,
I'm sorry that I'm late doing week 9 assignment, but it would be helpful if you give me some advice on my codes.
I got some problems as belows:
i) I don't exactly know how to generate code, which only affects 'the outcome' or 'the treatment'. The thing I only remember is 'rnormal(), but I'm not sure if it's making a specific affect to the outcome/treatment'
ii) Data is not combining well. I've set how to run regression with different models, but failed to combine results from the baseline model with different sample sizes.

Please, refer to my codes below:
------------------------------------------------------------------------------------------------------
/* Part 1-(1~2): Develop some data generating process for data X's and for outcome Y  */

/*i = District level   */
/*j = Village level  */
/*k = Entrepreneur level  */

set seed 1234 // Set a seed to create same data
clear
set obs 8 // Fix number of districts as '8'
gen district = _n
gen e_i = rnormal(0,3) // district effects

/* District level effects (i) */
gen large_dist = runiform()<0.05 // randomly assign large/small status on districts

/* Part 1-(3): Make Confounder  */

/* Village level effects (j)  */
expand 10 // create 10 villages in each districts
bysort district: gen village = _n // create village id
gen e_ij = rnormal(0,4) // village effects
bysort district: gen vill_pop = 1+int(10-1+1)*runiform() // Create a Continuous Covariate (Confounder!!!): Population of villages (unit: 1,000 people)

/* Entrepreneur level effects (k)  */
expand 25+int((40-25+1)*runiform()) // generate entrepreneur level dataset, each village will have 25 entrepreneurs
bysort district village: generate entrepreneur =_n //generate entrepreneur id
generate e_ijk = rnormal(0,5) //gen entrepreneur level effects

/* Part 1-(3): Make covariates (Only affect outcome, and only affect treatment)  */

	* Generate variable: Entrepreneur business size (entrepreneur level)
	generate rev_over_employ = runiform() 
	egen business_size = cut(rev_over_employ), at (0, 0.4, 0.7, 1) icodes 
	label define business_size 0 "Micro" 1 "Small_and_Medium" 2 "Big" 
	label values business_size business_size 
	tabulate business_size, generate(bsize) 

	/* Generage Covariate 1 (Only affects the outcome):  */
	gen export = rnormal(500, 150)
	
	/* Generage Covariate 2 (Only affects the treatment):  */
	gen tourist_influx = rnormal(200, 35)
	
/*DGP (Baseline model)  */
gen sales = 125 ///
	+ (1.4)*large_dist*runiform(-2,-1) ///
	+ 1.7*vill_pop ///
	+ e_i + e_ij + e_ijk
	
save part1_models_data, replace 

/* Part 1-(4): Run the 5 different regression with different sample sizes!!!  */

/* 
Reg 1: Baseline model 
gen sales = 125 + (1.4)*large_dist*runiform(-2,-1) + 1.7*vill_pop + e_i + e_ij + e_ijk

Reg 2: Baseline model + Business size effect
gen sales = 125 + (1.4)*large_dist*runiform(-2,-1) + 1.7*vill_pop + 1*bsize1 + 2*bsize2 + 5*bsize3 + e_i + e_ij + e_ijk

Reg 3: Baseline model + Business size effect + Export Effect
gen sales = 125 + (1.4)*large_dist*runiform(-2,-1) + 1.7*vill_pop + 1*bsize1 + 2*bsize2 + 5*bsize3 + 2.3*export + e_i + e_ij + e_ijk

Reg 4: Baseline model + Business size effect + Export Effect + Tourist effect
gen sales = 125 + (1.4)*large_dist*runiform(-2,-1) + 1.7*vill_pop + 1*bsize1 + 2*bsize2 + 5*bsize3 + 2.3*export + 1.4*tourist_influx + e_i + e_ij + e_ijk

Reg 5: Omitted variable model
gen sales = 125 + 1.7*vill_pop + 1*bsize1 + 2*bsize2 + 5*bsize3 + e_i + e_ij + e_ijk
*/ 

clear 

capture program drop normal_reg 
program define normal_reg_1, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part1_models_data.dta"
	sample `samplesize', count
	
	reg sales large_dist vill_pop e_i e_ij e_ijk
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

normal_reg_1, samplesize(100)
  /* Randomly sampled 100  */

/* Part 1-(4): Simulate program by running 500 times with different N  */

clear
tempfile part1_data
save `part1_data', replace emptyok

forvalues i=1/4{
	local ss = 10^`i'
	tempfile sims
	simulate N=r(N) beta_coef=r(beta) sem=r(sem) pvalues=r(pvalue) ci_l=r(ci_l) ci_u=r(ci_u), reps(500) seed(2023) saving(`sims'): normal_reg_1, samplesize(`ss')

use `sims', clear
append using `part1_data'
save `part1_data',replace
	
}

**Comment by genjuro86** [comment]
Hello,
I've tried to run regression as instruction.
However, I got stuck at some points as below.

i) I cannot find differences among sample sizes by graph. Thus, it makes me hard to understand what the assignment tries to teach.
Could you please help me how I can see the differences among different sample sizes? (Should I change numbers? or models?)

ii) It was difficult for me to produce a figure showing the mean and varianc of beta. What I learned was just 'histogram beta, by(N)' or 'histogram sem, by(N)'


Here is my codes below.

*** Part 1 ***

set seed 1234 
clear
set obs 8 
gen district = _n

expand 100

	gen bsize = rnormal()
	gen tourist_influx = rnormal()
	gen export = rnormal()
	gen treat = (bsize + tourist_influx + district/4 + rnormal()) > 0.6
	gen sales = (1.2)*bsize + export - district/20 + 1.5*treat + rnormal()
save part1_model, replace 

capture program drop normal_reg_1 
program define normal_reg_1, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part1_model.dta"
	sample `samplesize', count
	
	reg sales treat
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m1 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_2
program define normal_reg_2, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part1_model.dta"
	sample `samplesize', count
	
	reg sales treat bsize
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m2 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_3
program define normal_reg_3, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part1_model.dta"
	sample `samplesize', count
	
	reg sales treat bsize i.district
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m3 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_4
program define normal_reg_4, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part1_model.dta"
	sample `samplesize', count
	
	reg sales treat tourist_influx export 
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m4 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_5
program define normal_reg_5, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part1_model.dta"
	sample `samplesize', count
	
	reg sales treat bsize i.district tourist_influx export  
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m5 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

clear
tempfile combined
save `combined', replace emptyok

forvalues i=1/4{
    forvalues j=1/5{
	local ss = 10^`i'
	tempfile sims
	simulate N=r(N) beta_coef=r(beta) sem=r(sem) pvalues=r(pvalue) ci_l=r(ci_l) ci_u=r(ci_u), 	reps(500) seed(2023) saving(`sims'): normal_reg_`j', samplesize(`ss')

	use `sims', clear
	append using `combined'
	save `combined',replace
	}
}

use `combined', clear

histogram beta, by(N)



*** Part 2 ***

set seed 1234 
clear
set obs 8 
gen district = _n

expand 100

	gen bsize = rnormal()
	gen tourist_influx = rnormal()
	gen export = rnormal()
	gen treat = (bsize + tourist_influx + district/4 + rnormal()) > 0.6
	gen sales = (1.2)*bsize + export - district/20 + 1.5*treat + rnormal()

	gen channel = (2.5)*treat*runiform()
	gen collider = sales*treat + rnormal()
	
save part2_model, replace 

capture program drop normal_reg_1 
program define normal_reg_1, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part2_model.dta"
	sample `samplesize', count
	
	reg sales treat
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m1 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_2
program define normal_reg_2, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part2_model.dta"
	sample `samplesize', count
	
	reg sales treat bsize
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m2 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_3
program define normal_reg_3, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part2_model.dta"
	sample `samplesize', count
	
	reg sales treat bsize i.district
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m3 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_4
program define normal_reg_4, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part2_model.dta"
	sample `samplesize', count
	
	reg sales treat tourist_influx export 
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m4 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end

capture program drop normal_reg_5
program define normal_reg_5, rclass 
	syntax, samplesize(integer)
	clear
	use "C:\Genjuro\Georgetown_University\2_Spring_2023\Research_Design_and_Implementation\part2_model.dta"
	sample `samplesize', count
	
	reg sales treat bsize i.district tourist_influx export  
	
        matrix results = r(table)
	matrix list results
	
	return scalar N = e(N)
	return scalar m5 = _b[treat]
	return scalar beta = results[1,1]
	return scalar sem = results[2,1]
	return scalar pvalue = results[4,1]
	return scalar ci_l = results[5,1]
	return scalar ci_u = results[6,1]
 
end


clear
tempfile combined
save `combined', replace emptyok

forvalues i=1/4{
    forvalues j=1/5{
	local ss = 10^`i'
	tempfile sims
	simulate N=r(N) beta_coef=r(beta) sem=r(sem) pvalues=r(pvalue) ci_l=r(ci_l) ci_u=r(ci_u), reps(500) seed(2023) saving(`sims'): normal_reg_`j', samplesize(`ss')

	use `sims', clear
	append using `combined'
	save `combined',replace
	}
}

use `combined', clear

histogram beta, by(N)

**Comment by genjuro86** [comment]
Hello,
I've tried to make a model with Outcome(Y), treatment, confounder, covariates, and noises.
However, I got stuck at **how to divide values of treatment variable, which are "0" or "1", with equal observations.**

Total population is 500, so **I would like to 250 observations non-treated with program "0" (control group),
and the other 250 observations treated with program "1" (treatment group).**

It would be very helpful if you give us some tips to sovle it!
Here is my code below. 


clear
set obs 5
expand 100
gen x = _n

gen freetime = rnormal() // Confounder
gen mom_educ = rnormal() // Only affects outcome
gen work_exper = rnormal() // Only affects treatment
gen program = (freetime + (0.8)*work_exper + rnormal()) > 0.2  // Treatment
gen income = (2.3)*program + (1.3)*freetime + mom_educ/10 + rnormal() // Outcome

**Reply by hamzamajoka** [reply]
you can do something like this:

```
set seed 198924

clear
set obs 5
expand 100
gen x = _n

gen freetime = rnormal() // Confounder
gen mom_educ = rnormal() // Only affects outcome
gen work_exper = rnormal() // Only affects treatment
gen random_program = freetime + (0.8)*work_exper + rnormal()  //generate the underlying continuous var that will determine treatment status

summ random_program,d  //use summarize command to find the median (i.e. P50) 


gen program = random_program > r(p50) // use the median value so that Treatment is 1:1
tab program // this should give 250 in each arm
gen income = (2.3)*program + (1.3)*freetime + mom_educ/10 + rnormal() // Outcome
```

Let me know if this answers your question.

**Reply by genjuro86** [reply]
This is really helpful!! Thank you so much.
I couldn't think about dividing program by two steps.

**Comment by genjuro86** [comment]
Hello,
For week10, I could calculate the "powers" with different sample sizes between two models (biased and unbiased)
However, I got stuck at how to find the "minimum sample size".

To use "power twomeans" or "power twoproportions", I think we need to know two outcomes with and without treatment.

From this stage, I am not sure how to deduce numbers (ex. 500 550) to put after "power twomeans/twoproportions"

It would be very helpful if you give us some tips to sovle it!
Here is my code below.

capture program drop normal_reg
program define normal_reg, rclass
syntax, samplesize(integer)

clear
set obs 5
gen school = _n
expand `samplesize'

gen freetime = rnormal(11,3) // Confounder
gen mom_educ = rnormal(5,2) // Only affects outcome
gen work_exper = rnormal(3,0.5) // Only affects treatment
gen random_program = freetime + (0.8)*work_exper + rnormal() //Underlying continuous var that will determine treatment status

summ random_program, d  //Find the median (i.e. P50) 

gen program = random_program > r(p50) // Control and Treatment group will be 1:1
tab program 
gen income = (2.3)*program + (1.3)*freetime + mom_educ/10 + 6*rnormal() // Outcome

reg income program freetime mom_educ  // Unbiased reg model
matrix results = r(table)

return scalar N1= e(N)
return scalar beta1 = results[1,1]
return scalar pvalue1 = results[4,1]

reg income mom_educ freetime  // Biased reg model
matrix results = r(table)

return scalar N2= e(N)
return scalar beta2 = results[1,1]
return scalar pvalue2 = results[4,1]

end


clear 
tempfile combined sims
save `combined', replace emptyok

forvalues i=1/4 {
	local ss = 10^`i'
	simulate N=r(N1) beta1=r(beta1) beta2=r(beta2) pval1 = r(pvalue1) pval2 = r(pvalue2), rep(100) saving(`sims', replace): ///
	normal_reg, samplesize(`ss')

	use `sims', clear

	append using `combined'
	save `combined', replace

	}

	
use `combined', clear

gen sig1 = 0
replace sig1 = 1 if pval1 < 0.05 // This is for unbiased model

gen sig2 = 0
replace sig2 = 1 if pval2 < 0.05 // This is for biased model

sum sig1 sig2
bysort N: egen power_unbiased = mean(sig1)
bysort N: egen power_biased = mean(sig2)  // Calculate the "power"

graph box power_unbiased power_biased, by(N) yline(0.8) noout // By graph
